### PR TITLE
[Modules] New Microsoft.Compute/sshPublicKey

### PR DESCRIPTION
**Pull request by shawntmeyer**
This is a new module to generate a ssh Public Key. I cannot generate a status badge because I am unable to register the new workflow after resetting my main branch to the upstream commit, but the static tests and deployment tests worked successfully.

This resource is strange in that it really has no value. You cannot use this resource to generate an actual public key value. You must provide the public key string in the publicKey parameter if you want the resource to contain an actual public key.

The  common test uses a deployment script to generate a temporary ssh public key via the 'New-AzSshKey' cmdlet and grab the public key string from this object to provide it as the publicKey parameter value.

# Type of Change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] Update to documentation

# Checklist

- [X ] I'm sure there are no other open Pull Requests for the same update/change
- [X] My corresponding pipelines / checks run clean and green without any errors or warnings
- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (readme)
- [X] I did format my code

### CODE REVIEW

File: modules/Microsoft.Compute/sshPublicKeys/.test/.scripts/New-SshPubKey.ps1 [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿param(
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/.test/.scripts/New-SshPubKey.ps1:1]: Hey @shawntmeyer you could not have known, but please copy & use this [script](https://github.com/Azure/ResourceModules/blob/main/modules/Microsoft.Compute/virtualMachines/.test/.scripts/New-SSHKey.ps1).

To make matters easier in the future, we also have an [issue](#2619) open to centralize the used test utilities.

File: modules/Microsoft.Compute/sshPublicKeys/.test/common/dependencies.bicep [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            '${managedIdentity.id}': {}
+        }
+    }
+    properties: {
+        azPowerShellVersion: '8.0'
+        retentionInterval: 'P1D'
+        arguments: '-resourceGroupName ${resourceGroup().name} -keyName ${tempKeyName}'
+        scriptContent: loadTextContent('../.scripts/New-SshPubKey.ps1')
+        cleanupPreference: 'OnExpiration'
+        forceUpdateTag: utcValue
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/.test/common/dependencies.bicep:47]: Is this needed? In our [Linux VM](https://github.com/Azure/ResourceModules/blob/main/modules/Microsoft.Compute/virtualMachines/.test/linux.min/dependencies.bicep#:~:text=%7D-,resource%20sshDeploymentScript%20%27Microsoft.Resources/deploymentScripts%402020%2D10%2D01%27%20%3D%20%7B,%7D,-resource%20sshKey%20%27Microsoft) implementation (that uses SSH keys) we didn't use the property. That naturally, if it makes sense to provide it, we can add it there.

File: modules/Microsoft.Compute/sshPublicKeys/.test/common/dependencies.bicep [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    dependsOn: [
+        msi_ContributorRoleAssignment
+    ]
+}
+
+@description('The public key to be added to the SSH Public Key resource.')
+output publicKey string = createPubKeyScript.properties.outputs.pubKey
+
+@description('The Resource ID of the managed Identity')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/.test/common/dependencies.bicep:57]:
```suggestion
@description('The resource ID of the managed Identity')
```

File: modules/Microsoft.Compute/sshPublicKeys/.test/common/deploy.test.bicep [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.compute.sshPublicKeys-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'sshcom'
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/.test/common/deploy.test.bicep:15]:
```suggestion
param serviceShort string = 'cspkcom'
```

File: modules/Microsoft.Compute/sshPublicKeys/.test/common/deploy.test.bicep [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  name: resourceGroupName
+  location: location
+}
+
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-nestedDependencies'
+  params: {
+    generateSshPubKeyScriptName: 'dep-<<namePrefix>>-ds-${serviceShort}-generateSshPubKey'
+    tempKeyName: '${uniqueString(resourceGroup.name, baseTime)}-temp'
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/.test/common/deploy.test.bicep:39]: Why not reuse the same name like
`sshKeyName: 'dep-<<namePrefix>>-ssh-${serviceShort}'`

File: modules/Microsoft.Compute/sshPublicKeys/.test/min/deploy.test.bicep [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.compute.sshPublicKeys-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'sshmin'
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/.test/min/deploy.test.bicep:15]:
```suggestion
param serviceShort string = 'cspkmin'
```

File: modules/Microsoft.Compute/sshPublicKeys/deploy.bicep [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    principalIds: roleAssignment.principalIds
+    principalType: contains(roleAssignment, 'principalType') ? roleAssignment.principalType : ''
+    roleDefinitionIdOrName: roleAssignment.roleDefinitionIdOrName
+    condition: contains(roleAssignment, 'condition') ? roleAssignment.condition : ''
+    delegatedManagedIdentityResourceId: contains(roleAssignment, 'delegatedManagedIdentityResourceId') ? roleAssignment.delegatedManagedIdentityResourceId : ''
+    resourceId: sshPublicKey.id
+  }
+}]
+
+@description('The resource group the  disk was deployed into.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/deploy.bicep:70]:
```suggestion
@description('The name of the Resource Group the SSH Public Key was created in.')
```

File: modules/Microsoft.Compute/sshPublicKeys/deploy.bicep [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+output resourceGroupName string = resourceGroup().name
+
+@description('The resource ID of the disk.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/deploy.bicep:73]:
```suggestion
@description('The resource ID of the Public SSH Key.')
```

File: modules/Microsoft.Compute/sshPublicKeys/deploy.bicep [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+output resourceId string = sshPublicKey.id
+
+@description('The name of the disk.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/deploy.bicep:76]:
```suggestion
@description('The name of the Public SSH Key.')
```

File: modules/Microsoft.Compute/sshPublicKeys/readme.md [hunk 10]
```diff
@@ -0,0 +1,253 @@
+# Compute Disks `[Microsoft.Compute/sshPublicKeys]`
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/readme.md:1]:
```suggestion
# Public SSH Keys `[Microsoft.Compute/sshPublicKeys]`
```

File: modules/Microsoft.Compute/sshPublicKeys/readme.md [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+This template deploys a disk
```
Comment (AlexanderSehr) [inline modules/Microsoft.Compute/sshPublicKeys/readme.md:3]:
```suggestion
This template deploys a Public SSH Key resource.
> Note: It does not generate the key for you but acts as a store for a provided key. The key can be generated using for example a PowerShell script in a Deployment Script.

```

Comment (shawntmeyer) [reply modules/Microsoft.Compute/sshPublicKeys/readme.md:3]: Does the invoke command output a private key? the use case here is to generate the key pair, the sshpublickey holds the public key property which is used when deploying the VM. The private key is stored locally in a file .\ssh\filename and we get that private key and store it in keyvault. it is used for sshlogon to Linux.

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**